### ISSUE TITLE
Update to Glutin 0.21.0

### ISSUE THREAD

**Issue by ZeGentzy** [servo/servo#23189]
https://github.com/servo/servo/blob/7b8a898ac2ffef5ced8202dfd8bb40e37b972334/ports/servo/Cargo.toml#L49

Hey folks. Glutin v0.21.0 is out.

**Comment by paulrouget** [comment]
There's some non-trivial changes in this version of Glutin (they way the windows are built and the way the Context is carried, at least, didn't go very far). Testing this will require a bit of work.

@ZeGentzy are you planning to release 0.21 soon?

**Comment by ZeGentzy** [comment]
Planning to give it two or so weeks, to do other stuff on my TODO list.

**Comment by paulrouget** [comment]
First issue: glutin_gles2_sys needs a version bump to include the gl_generator update. Otherwise we get 2 different version of gl_generator.

**Comment by paulrouget** [comment]
In general, compiling something with glutin should not generate a Cargo.lock with 2 version of `gl_generator`.

**Comment by paulrouget** [comment]
Here are a few changes I had to make to avoid dependencies on different version of same packages:

glutin:

```diff
diff --git a/glutin/Cargo.toml b/glutin/Cargo.toml
index 4f4c9563..a1b13760 100644
--- a/glutin/Cargo.toml
+++ b/glutin/Cargo.toml
@@ -56,7 +56,7 @@ parking_lot = "0.7"
 
 [target.'cfg(any(target_os = "linux", target_os = "freebsd", target_os = "dragonfly", target_os = "netbsd", target_os = "openbsd"))'.dependencies]
 osmesa-sys = "0.1"
-wayland-client = { version = "0.21", features = ["egl", "dlopen"] }
+wayland-client = { version = "0.23", features = ["egl", "dlopen"] }
 libloading = "0.5"
 glutin_egl_sys = { version = "0.1.2", path = "../glutin_egl_sys" }
 glutin_glx_sys = { version = "0.1.3", path = "../glutin_glx_sys" }
diff --git a/glutin_egl_sys/Cargo.toml b/glutin_egl_sys/Cargo.toml
index 301268a6..2f5d9f73 100644
--- a/glutin_egl_sys/Cargo.toml
+++ b/glutin_egl_sys/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "glutin_egl_sys"
-version = "0.1.2"
+version = "0.1.3"
 authors = ["The glutin contributors", "Hal Gentz <zegentzy@protonmail.com>"]
 description = "The egl bindings for glutin"
 repository = "https://github.com/tomaka/glutin"
diff --git a/glutin_gles2_sys/Cargo.toml b/glutin_gles2_sys/Cargo.toml
index 0d8fa0b2..502ec104 100644
--- a/glutin_gles2_sys/Cargo.toml
+++ b/glutin_gles2_sys/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "glutin_gles2_sys"
-version = "0.1.2"
+version = "0.1.3"
 authors = ["The glutin contributors", "Hal Gentz <zegentzy@protonmail.com>"]
 description = "The gles2 bindings for glutin"
 repository = "https://github.com/tomaka/glutin"
diff --git a/glutin_glx_sys/Cargo.toml b/glutin_glx_sys/Cargo.toml
index 050faf4b..73a1ac80 100644
--- a/glutin_glx_sys/Cargo.toml
+++ b/glutin_glx_sys/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "glutin_glx_sys"
-version = "0.1.3"
+version = "0.1.4"
 authors = ["The glutin contributors", "Hal Gentz <zegentzy@protonmail.com>"]
 description = "The glx bindings for glutin"
 repository = "https://github.com/tomaka/glutin"
diff --git a/glutin_wgl_sys/Cargo.toml b/glutin_wgl_sys/Cargo.toml
index cabfbab3..90affc6a 100644
--- a/glutin_wgl_sys/Cargo.toml
+++ b/glutin_wgl_sys/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "glutin_wgl_sys"
-version = "0.1.2"
+version = "0.1.3"
 authors = ["The glutin contributors", "Hal Gentz <zegentzy@protonmail.com>"]
 description = "The wgl bindings for glutin"
 repository = "https://github.com/tomaka/glutin"
```

```diff
diff --git a/Cargo.toml b/Cargo.toml
index 35fde6d6..fcbd2a5d 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -63,8 +63,8 @@ features = [
 ]
 
 [target.'cfg(any(target_os = "linux", target_os = "dragonfly", target_os = "freebsd", target_os = "openbsd", target_os = "netbsd"))'.dependencies]
-wayland-client = { version = "0.21", features = [ "dlopen", "egl", "cursor"] }
-smithay-client-toolkit = "0.4.3"
+wayland-client = { version = "0.23", features = [ "dlopen", "egl", "cursor"] }
+smithay-client-toolkit = "0.6"
 x11-dl = "2.18.3"
 parking_lot = "0.7"
 percent-encoding = "1.0"
```

**Comment by paulrouget** [comment]
I manage to test the rc version. Beside the above issues, it looks like it's work ok.

The new way to hold a context is a bit annoying though (`Context<T>` which keeps switching between current and not current, and `make_current` consuming self).

- Servo: https://github.com/paulrouget/servo/tree/glutin-update-rc
- rust-webvr: https://github.com/paulrouget/rust-webvr/tree/glutin-update-rc
- skia: https://github.com/paulrouget/skia/tree/glutin-update-rc
- glx: https://github.com/paulrouget/rust-glx/tree/glutin-update-rc
- offscreen-gl-context: https://github.com/paulrouget/rust-offscreen-rendering-context/tree/glutin-update-rc

**Comment by ZeGentzy** [comment]
I've submitted your patches, @paulrouget, thank you very much.

**Comment by paulrouget** [comment]
> I remembered that updating wayland-client for winit has already done, on their evlp2 branch. I think they're waiting on an X11 impl and a MacOS impl of the new evlp2.

This means we have a duplicate package (nix). I think we can work around that for now.

**Comment by jdm** [comment]
This was reverted in https://github.com/servo/servo/pull/23442.

**Comment by jdm** [comment]
This merged in #23457.
